--- patch-2.5.4/patch.c.suffix	Thu Nov  2 17:57:46 2000
+++ patch-2.5.4/patch.c	Thu Nov  2 17:58:25 2000
@@ -145,8 +145,7 @@
 		 : posixly_correct - 1);
 
     val = getenv ("SIMPLE_BACKUP_SUFFIX");
-    if (val && *val)
-      simple_backup_suffix = val;
+    simple_backup_suffix = val && *val ? val : ".orig";
 
     if ((version_control = getenv ("PATCH_VERSION_CONTROL")))
       version_control_context = "$PATCH_VERSION_CONTROL";
